### PR TITLE
Only use rich text value internally to the RichText component

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

This is a proposal to use the string format on the RichText component by default.

Benefits:

* The block attribute values are plain HTML, so I'd be no longer concerned about plugin authors manipulating the raw rich text value directly.
* Plain HTML can be passed to `createBlock` to create templates.
* The rich text value is only used internally to rich text. No other code should be aware of this value shape.
* It would be okay if the value is not serialisable as JSON. It will never be in the global state.
* The value can be stored in the block comments.
* No need to start recommending different formats for different use cases such as post meta.
* The format is internal only so we can continue to make improvements.

Cons:

* Slightly less performant, but I would consider this is negligible.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation.

### CODE REVIEW

**Comment by iseulde** [comment]
@youknowriad This may be of interest to you.

File: packages/blocks/src/api/matchers.js
Comment (youknowriad) [inline packages/blocks/src/api/matchers.js]: Why the regular `html` matcher from `hpq` is not sufficient?

**Review by youknowriad — COMMENTED** [review]
It does feel nicer from the block's author pov to only have to deal with `html` strings. (I proposed this a long time ago :P) But what about performance? Having to serialize/parse html for RichText on each render seems overkill?

File: packages/blocks/src/api/matchers.js
Comment (iseulde) [reply packages/blocks/src/api/matchers.js]: Because this one would accept `multilineTag` as an argument for `multiline` RichText values. Otherwise we'd have to go back to the `query` matcher and map the values to and from `RichText`. I don't mind doing that with strings though. Before this was an issue because it would create a new array or object every time, but with a string we can easily compare.

**Comment by iseulde** [comment]
> But what about performance? Having to serialize/parse html for RichText on each render seems overkill?

Honestly I think it is negligible, but we can do some testing. The only addition would be serialising to HTML on input and this should be very fast. Additionally this only happens per RichText instance, not the whole block or whole editor content.

File: packages/blocks/src/api/matchers.js
Comment (youknowriad) [reply packages/blocks/src/api/matchers.js]: Thanks for explaining, I'm not too concerned about that. `multiline` seems fine.

**Comment by youknowriad** [comment]
I'm still hesitant between this and #10370. Regardless, I think we should do something in `synchronizeBlocksWithTemplate` in this PR to ensure a smooth transition from old templates where the source is `html` but the template contain an `array` as a value. (upgrade and probably deprecate as well).

The last hesitation for me would be the performance, so we do when do we do this parsing/serialization:

 - on first render of the RichText component we parse
 - if the prop value changes, we parse again?
 - When tinymce triggers a change, I suppose we parse and serialize?

How does this differ from having the attribute as the internal representation directly?

Also pinging @aduth for a second opinion.

**Comment by youknowriad** [comment]
I pushed a commit on this branch making the current templates work:

```
add_action( 'init', function() {
	register_post_type( 'book', [
		'label' => 'Book',
		'show_in_rest' => true,
		'public' => true,
		'show_ui' => true,
		'supports' => [ 'title', 'editor' ],
		'template' => [
			[ 'core/paragraph', [
				'content' => 'Hello first paragraph!',
			] ],
			[ 'core/paragraph', [
				'content' => [
					'Hello ',
					[
						'type' => 'strong',
						'props' => [ 'children' => 'world' ]
					],
					'!'
				],
			] ],

			[ 'core/gallery', [
				'images' => [
					[
						'url' => 'https://picsum.photos/200/300',
						'caption' => [
							'Hello ',
								[
									'type' => 'strong',
									'props' => [ 'children' => 'world' ]
								],
								'!'
						]
					]
				],
			] ],
		],
		'template_lock' => 'all',
	] );
} );
```

**Comment by youknowriad** [comment]
Last thing here, I'd like to add is to remove the "children" deprecations for now. My idea is the following:

 - Ship this new format without deprecations
 - This will allow us to test it in real scenarios in prod without asking people to change their blocks yet.
 - If the result is satisfying in terms of performance, we reintroduce the deprecations in the next version asking people to use "html strings" consistently.

File: test/e2e/specs/splitting-merging.test.js
Comment (gziolo) [inline test/e2e/specs/splitting-merging.test.js]: That's surprising, any ideas why it needs to be updated?

File: test/e2e/specs/__snapshots__/writing-flow.test.js.snap
Comment (gziolo) [inline test/e2e/specs/__snapshots__/writing-flow.test.js.snap]: Do we know why it's different?

File: test/e2e/specs/__snapshots__/templates.test.js.snap
Comment (gziolo) [inline test/e2e/specs/__snapshots__/templates.test.js.snap]: Do we know why it no longer prints `<p />`?

File: packages/editor/src/components/rich-text/index.native.js
Comment (gziolo) [inline packages/editor/src/components/rich-text/index.native.js]: @hypest can you help testing if this still works with mobile app?

File: packages/editor/src/components/rich-text/index.js
Comment (gziolo) [inline packages/editor/src/components/rich-text/index.js]: It feels like `RichText` component should be moved to `@wordpress/rich-text` package to avoid this duality where you can have two version of each of the public methods:
```js
import { join } from '@wordpress/rich-text';
import { RichText } from '@wordpress/editor';

join === RichText.join; // false
```

It feels like with the changes proposed `join` should remain private, and RichText.join would be the only public API.

**Review by gziolo — COMMENTED** [review]
There are many subtle differences in the output html produced for snapshots in unit tests which raises questions why it has changed.

File: packages/block-library/src/paragraph/index.js
Comment (aduth) [inline packages/block-library/src/paragraph/index.js]: What's the difference between this and `attributes.content + attributesToMerge.content` ?

Will that difference be obvious to someone who thinks that they are working with strings?

File: packages/editor/src/components/rich-text/index.native.js
Comment (hypest) [reply packages/editor/src/components/rich-text/index.native.js]: Tested against the current mobile app master (https://github.com/wordpress-mobile/gutenberg-mobile/commit/6f5e3507ff40eb351ef7c5e4c16a9db1cdae2a07) and works fine 👍

File: packages/blocks/src/api/matchers.js
Comment (aduth) [reply packages/blocks/src/api/matchers.js]: This feels a bit strange to me as well, since it's something of a preprocessing filter that doesn't serve much purpose outside the RichText use-case.

Curious: What happens if we don't have this at all?

File: packages/blocks/src/api/parser.js
Comment (aduth) [inline packages/blocks/src/api/parser.js]: It's not obvious what purpose `multiline` serves when considering `html` in isolation outside the RichText use-case. We're kinda muddying the distinction between these sources.

File: packages/editor/src/components/rich-text/index.js
Comment (aduth) [inline packages/editor/src/components/rich-text/index.js]: Are these considered deprecated? Will they ever be removed?

**Comment by aduth** [comment]
Previously: #5380 (Some "gotchas" which might be relevant here)

File: packages/block-library/src/paragraph/index.js
Comment (iseulde) [reply packages/block-library/src/paragraph/index.js]: I'm actually changing this right now.

**Comment by aduth** [comment]
Reiterating Riad's point, this is certainly an improvement from the developer experience angle over what we'd be introducing with `rich-text` (and had with `children`). I feel quite positive about it from that angle.

Trying to understand the performance impact, I think the impact is an acceptable compromise. It makes me wonder what options we might have for future optimization: If I press a single character, shouldn't this be able to be applied to the internal representation easily without re-parsing/filtering the element in its entirety? Not saying that needs to be done here or even immediately after, but if that's an idea which has been considered to be pursued, I could see how it would offset (or even produce net benefit to) the performance cost paid here.

Re: #5380 and with regard to other ongoing work with annotations, do we plan to support the use case where developers can interact with the rich text value if they need to? I suppose that is a matter of using the `rich-text` package and its utilities? There is an added cost here if it requires double-creating the same `rich-text` value from a single field.

Interestingly one of the issues with #5380 which led to its closing was the need for a custom serializer, which we now have (twice over).

In all, I think we're in a much better position than we were in #5380 with our separations (`rich-text`, `escape-html`) to support this.

File: docs/reference/deprecated.md
Comment (iseulde) [inline docs/reference/deprecated.md]: There's a trailing 's' here :)

File: packages/editor/src/components/rich-text/index.js
Comment (iseulde) [reply packages/editor/src/components/rich-text/index.js]: We can remove multiline as it was added after the last release.

File: test/e2e/specs/__snapshots__/templates.test.js.snap
Comment (iseulde) [reply test/e2e/specs/__snapshots__/templates.test.js.snap]: Because the initial content of quote will be undefined. This is how it was before the rich text value merge:

https://github.com/WordPress/gutenberg/pull/7890/files#diff-632d636fb9c39f4f593363386f77cedf

File: packages/editor/src/components/rich-text/index.js
Comment (iseulde) [reply packages/editor/src/components/rich-text/index.js]: This has changed now.

File: packages/block-library/src/paragraph/index.js
Comment (iseulde) [reply packages/block-library/src/paragraph/index.js]: Updated.

File: packages/block-library/src/file/index.js
Comment (aduth) [inline packages/block-library/src/file/index.js]: `fileName` is a string. Why should we need utility.

**Review by aduth — CHANGES_REQUESTED** [review]
1. New post
2. Enter title
3. Click writing prompt
4. Cmd+B to bold
5. Click "Save draft"
6. Reload page

Expected: One of either:
- Paragraph block with empty strong tag
- No post content (per consistent with #9808)

Actual: Empty paragraph, no `strong`.

Related to writing flow E2E change.

File: packages/blocks/src/api/matchers.js
Comment (iseulde) [reply packages/blocks/src/api/matchers.js]: It would include e.g. the cite. As I said, the alternative would be to revert to using query and map the values on each render. I like that solution less, but since it's just a string now I don't care as much.

**Review by aduth — DISMISSED** [review]
1. New post
2. Click writing prompt
3. Cmd+B to bold
4. Type something
5. Press Enter
6. Press Backspace
7. Press Enter

Expected: Typing in second paragraph.
Actual: Second paragraph created, but caret remains in first.

Works as expected in 3.9.0.
Should have E2E case (could extend "should merge into inline boundary position")

File: packages/block-library/src/file/index.js
Comment (iseulde) [reply packages/block-library/src/file/index.js]: Yes, we could get rid of the is empty utility as well now. I'm fine either way.

Comment (aduth) [reply packages/block-library/src/file/index.js]: Considering them as strings, I see no need for the utility.

File: packages/blocks/src/api/matchers.js
Comment (aduth) [reply packages/blocks/src/api/matchers.js]: We're looking at this too much through the lens of Rich Text. If we need the matcher to support filtering, that's one thing. But on that basis alone, "multiline" doesn't maker much sense to me. I'd rather some general support of aggregates, a property which more accurately reflects the child selector (e.g. https://github.com/WordPress/gutenberg/pull/6716#issuecomment-388907201), or reverting to query as you mention.

**Comment by iseulde** [comment]
Thanks for the tests @aduth, addressing them now.

File: test/e2e/specs/splitting-merging.test.js
Comment (aduth) [reply test/e2e/specs/splitting-merging.test.js]: I don't know that its strictly an issue here, but caret placement in and around inline boundaries has been historically problematic and inconsistent and we should be careful to avoid worsening this. e.g. in some cases depending on internal state of TinyMCE, caret will still move inside inline boundary when backspacing (e.g. if Enter + Backspace when at end of a paragraph inside inline boundary).

Related: #9040 (internal state not an issue if its not preserved)

**Review by youknowriad — APPROVED** [review]
LGTM 👍 

I acknowledge that there's still a small bug when merging blocks. Let's create an issue for it and address it separately.

File: packages/editor/src/components/rich-text/index.native.js
Comment (mcsf) [reply packages/editor/src/components/rich-text/index.native.js]: ❤️ to both of you.

File: packages/blocks/src/api/matchers.js
Comment (aduth) [reply packages/blocks/src/api/matchers.js]: Where will this discussion continue? Are we going to launch 4.0 with the `multiline` property for `html` attribute sources?

Comment (youknowriad) [reply packages/blocks/src/api/matchers.js]: I personally thought it was a good compromise as opposed to the "artificial" key we had to use when using the `query` matcher. And I thought this "multiline" option was supposed to fix that.

**Comment by youknowriad** [comment]
@ktmn Sorry for the breakage. It was meant to be backward compatible but we probably missed something there, a bug. Can I suggest that you you try Gutenberg RC's before their release. We did these RC periods specifically to gather feedback about things we missed?

That said, can you create a separate issue with a small block example and the corresponding template that broke so we could take a look?

**Comment by ktmn** [comment]
@youknowriad Yeah it's my bad for missing the RC, never followed the WordPress blog to see it's announcement. 

I'm just trying to figure out what the attribute type for rich text needs to be and how to do the new templates.

> That said, can you create a separate issue with a small block example and the corresponding template that broke so we could take a look?

For example [this plugin](https://wordpress.org/plugins/advanced-gutenberg/) with Gutenberg 4.0, create an "Advanced button", add some text to it, save and refresh. `This block contains unexpected or invalid content.` This isn't my plugin or anything but it already exhibits some lack of backwards compatibility.

**Comment by youknowriad** [comment]
So here's the best path forward in term of blocks:

 - use string
 - use `html` instead of `children` as a source
 - use strings in templates.

We did manage to try to transform array in templates automatically to strings when the source is `html` to ensure backward compatibility.

How's that look for you?

**Comment by youknowriad** [comment]
What I am wondering is why your blocks without any change broke the templates as we tried to keep the `children` sources with their corresponding templates as is.

**Comment by ktmn** [comment]
> What I am wondering is why your blocks without any change broke the templates as we tried to keep the children sources with their corresponding templates as is.

In short, this is how my block looked like:

```js
{
	attributes: {
		content: {
			type: 'array',
		},
	},
	edit: ({attributes, setAttributes}) => {
		return (
			<div className='custom-block'>
				<wp.editor.RichText
					value={attributes.content}
					onChange={value => setAttributes({content: value})}
				/>
			</div>
		)
	},
	save: ({attributes}) => {
		return (
			<div className='custom-block'>
				{attributes.content}
			</div>
		)
	},
}
```

After 4.0.0 that produced `This block contains unexpected or invalid content.` after refreshing the editor, since the rich text now returns a string and I guess it get's removed from the block comments because it doesn't match the type specified.

So I changed it to string:

```js
{
	attributes: {
		content: {
			type: 'string',
		},
	},
	...
}
```

and ran into issues with html being escaped, such as bold text or multiline.
I figured out I gotta use `<RichText.Content>` so I did.

```js
{
	attributes: {
		content: {
			type: 'array',
		},
	},
	edit: ({attributes, setAttributes}) => {
		return (
			<div className='custom-block'>
				<wp.editor.RichText
					value={attributes.content}
					onChange={value => setAttributes({content: value})}
				/>
			</div>
		)
	},
	save: ({attributes}) => {
		return (
			<div className='custom-block'>
				<wp.editor.RichText.Content value={attributes.content} />
			</div>
		)
	},
}
```

Then I have a block that is basically a preset of blocks, so it looked something like this:

```js
{
	attributes: {},
	edit: ({attributes, setAttributes}) => {
		return (
			<div className='my-other-block'>
				<wp.editor.InnerBlocks
					template={[
						['custom/block', {
							content: ['Content'],
						}],
					]}
				/>
			</div>
		)
	},
	save: ({attributes}) => {
		return (
			<div className='my-other-block'>
				<wp.editor.InnerBlocks.Content />
			</div>
		)
	},
}
```

Notice the content in the template is type `array`, the old format, but the block earlier used `string` value now. It will render the `InnerBlocks` fine when you insert the block, but after refresh the content attribute is gone from the `custom/block` block comments and I get an error (~`The block has been modified externally`).

So to solve that I removed the type from `custom/block`:

```js
{
	attributes: {
		content: {},
	},
	...
}
```

Now it works. It can use string from `<RichText>` and array from old templates. That's how I reconciled 4.0.

I'm not saying anything is bugged, this is just how it went down for me.

Now I'm just wondering: 

1. Is it correct to have no type for the attribute?

2. Do I keep the old array templates (of which I have about 20 different ones, 1200 lines)? Or do I need to convert them to string format.

3. And if so, how does bold and `<br>` etc look like in the new string format? Are there examples?

**Comment by youknowriad** [comment]
> Is it correct to have no type for the attribute?

It's correct but I expect the first issue is that the block was initially doing something wrong: You're rendering the content in html and at the same time saving it to the block's wrapping comment (because there's no source).

You could try adding `source: "children", selector: ".custom-block"` to the attribute's definition.

**Comment by ktmn** [comment]
> You're rendering the content in html and at the same time saving it to the block's wrapping comment (because there's no source).

> You could try adding `source: "children", selector: ".custom-block"` to the attribute's definition.

I do that because the selectors are not capable handling more complex blocks.

For example if I have a button:

```js
<div class="button">
    Content
</div>
```

it can work easily with this attribute:

```js
{
    content: {
        source: 'html', // not 'children' anymore, right?
        selector: '.button',
    },
}
```

but when there's a block control that allows to add an icon for example, making the output like this:


```js
<div class="button">
    Content
    <i class="icon banana"></i>
</div>
```

I can no longer fetch only the content without having to wrap it first in a `span` or something.

Now maybe the button also has a toggle control allowing the button to *only* show the icon, resulting in this render:


```js
<div class="button">
    <i class="icon banana"></i>
</div>
```

then the content would be gone, even if the user just tried to test what this option looks like.

Of course I use the `source` when possible but most of the time I need to keep the value in the comments. I really hope that's supported, I mean it has been working fine always.

For string rich text `<em>` is stored in the comments like this `\u003cem\u003Content\u003c/em\u003e` and it works fine, but if I put this string in a template it doesn't. So how would I do this in templates?

**Comment by youknowriad** [comment]
There's something we need to check separately. An issue with this block would be good to check why the template is not working.

```
{
	attributes: {
		content: {
			type: 'array',
		},
	},
	edit: ({attributes, setAttributes}) => {
		return (
			<div className='custom-block'>
				<wp.editor.RichText
					value={attributes.content}
					onChange={value => setAttributes({content: value})}
				/>
			</div>
		)
	},
	save: ({attributes}) => {
		return (
			<div className='custom-block'>
				{attributes.content}
			</div>
		)
	},
}
```

---
Separately, I'd advocate to transform everything into string including the template content.

**Pull request merged**
